<style>
  body {
    color: black;
  }
  h1 {
    background-color: transparent;
    color: LightSteelBlue;
  }
  h3 {
    background-color: transparent;
    color: WhiteSmoke;
  }
  b {
    background-color: transparent;
    color: LightSteelBlue;
  }
</style>
<h1>Introduction</h1>
<h4>Column Names</h4>
<ul>
  <li><b>ArithmeticErrortarget:</b> the polarity of the tweet (0 = negative, 2 = neutral, 4 = positive)</li>
  <li><b>id:</b> The id of the tweet (2087)</li>
  <li><b>date:</b> The date of the tweet (Sat May 16 23:58:44 UTC 2009)</li>
  <li><b>flag:</b> The query (lyx). If there is no query, then this value is NO_QUERY.</li>
  <li><b>user:</b> The user that tweeted (robotickilldozr)</li>
  <li><b>text:</b> The text of the tweet (Lyx is cool)</li>
</ul>

<h4>Import</h4>

In [220]:
import csv
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import display
from datetime import datetime

if __name__ == '__name__':
    print('Starting Spark Session. All data is loaded into memory')

<h4>Read the File</h4>

Read the csv file into a Python dataframe and then write it into a new file

In [221]:
df = pd.read_csv('.databases/tweets.csv', encoding='ISO-8859-1')
display(df)
display(df.shape)

0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY   
0        0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY  \
1        0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2        0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3        0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4        0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   
...     ..         ...                           ...       ...   
1599994  4  2193601966  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599995  4  2193601969  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599996  4  2193601991  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599997  4  2193602064  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599998  4  2193602129  Tue Jun 16 08:40:50 PDT 2009  NO_QUERY   

         _TheSpecialOne_   
0          scotthamilton  \
1               mattycus   
2                ElleCTF   
3                 Karoli   
4               joy_wolf   
...                  ...   
1599994  AmandaMarie1028   
1599995      TheWDBoards   
1599996           bpbabe   
1599997     tinydiamondz   
1599998   RyanTrevMorris   

        @switchfoot http://twitpic.com/2y1zl - Awww, that's a bummer.  You shoulda got David Carr of Third Day to do it. ;D  
0           is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                                 @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                           my whole body feels itchy and like its on fire   
3           @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                             @Kwesidei not the whole crew   
...                                                                                                                     ...  
1599994                                                            Just woke up. Having no school is the best feeling ever   
1599995                                      TheWDB.com - Very cool to hear old Walt interviews!  ♫ http://blip.fm/~8bmta  
1599996                                                           Are you ready for your MoJo Makeover? Ask me for details   
1599997                                                   Happy 38th Birthday to my boo of alll time!!! Tupac Amaru Shakur   
1599998                                                      happy #charitytuesday @theNSPCC @SparksCharity @SpeakingUpH4H   

[1599999 rows x 6 columns]

(1599999, 6)

<style>
  body {
    color: black;
  }
  h1 {
    background-color: transparent;
    color: LightSteelBlue;
  }
  h3 {
    background-color: transparent;
    color: WhiteSmoke;
  }
  b {
    background-color: transparent;
    color: LightSteelBlue;
  }
</style>
<h1>Action</h1>

Get the basic information about the content of the file

In [222]:
df.columns = ['sentiment', 'id', 'date', 'flag', 'user', 'text']
print('The first five rows of the database')
display(df.head())
print('Information')
display(df.info())
print('Description')
df.describe()

The first five rows of the database


sentiment          id                          date      flag   
0          0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY  \
1          0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2          0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3          0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4          0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   

            user   
0  scotthamilton  \
1       mattycus   
2        ElleCTF   
3         Karoli   
4       joy_wolf   

                                                                                                              text  
0  is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                        @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                  my whole body feels itchy and like its on fire   
3  @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                    @Kwesidei not the whole crew

Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599999 entries, 0 to 1599998
Data columns (total 6 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   sentiment  1599999 non-null  int64 
 1   id         1599999 non-null  int64 
 2   date       1599999 non-null  object
 3   flag       1599999 non-null  object
 4   user       1599999 non-null  object
 5   text       1599999 non-null  object
dtypes: int64(2), object(4)
memory usage: 73.2+ MB


None

Description


sentiment            id
count  1.599999e+06  1.599999e+06
mean   2.000001e+00  1.998818e+09
std    2.000001e+00  1.935757e+08
min    0.000000e+00  1.467811e+09
25%    0.000000e+00  1.956916e+09
50%    4.000000e+00  2.002102e+09
75%    4.000000e+00  2.177059e+09
max    4.000000e+00  2.329206e+09

<h3>Point *</h3>Now let us get the number of unique values in each column; we had 

**df.unstack().groupby(level=0).nunique()**

as a suggestion but it was over 6 times slower

In [223]:
# we had 
# df.unstack().groupby(level=0).nunique()
# as a suggestion but it was over 6 times slower
df.apply(pd.Series.nunique)

sentiment          2
id           1598314
date          774362
flag               1
user          659775
text         1581465
dtype: int64

In [224]:
for _ in df.columns:
    display(_)
    display(df[_].value_counts())

'sentiment'

sentiment
4    800000
0    799999
Name: count, dtype: int64

'id'

id
2190457769    2
1974742852    2
2062516845    2
1551586713    2
1563681287    2
             ..
2197311343    1
2197311196    1
2197311146    1
2197310899    1
2193602129    1
Name: count, Length: 1598314, dtype: int64

'date'

date
Mon Jun 15 12:53:14 PDT 2009    20
Fri May 29 13:40:04 PDT 2009    17
Mon Jun 15 13:39:50 PDT 2009    17
Fri May 22 05:10:17 PDT 2009    17
Fri Jun 05 11:05:33 PDT 2009    16
                                ..
Sun Jun 07 12:36:09 PDT 2009     1
Sun Jun 07 12:36:07 PDT 2009     1
Sun Jun 07 12:36:04 PDT 2009     1
Sun Jun 07 12:36:03 PDT 2009     1
Tue Jun 16 08:40:50 PDT 2009     1
Name: count, Length: 774362, dtype: int64

'flag'

flag
NO_QUERY    1599999
Name: count, dtype: int64

'user'

user
lost_dog           549
webwoke            345
tweetpet           310
SallytheShizzle    281
VioletsCRUK        279
                  ... 
iheartrobpattz       1
67trinity            1
Sibby                1
mAnyA_15             1
bpbabe               1
Name: count, Length: 659775, dtype: int64

'text'

text
isPlayer Has Died! Sorry                                                                              210
good morning                                                                                          118
headache                                                                                              115
Good morning                                                                                          112
Headache                                                                                              106
                                                                                                     ... 
braces  tell me it will be okay...                                                                      1
is stuck at home without curry                                                                          1
@mrsduryee I've applied to about 70 since I lost my job in March...it certainly FEELS like a lot!       1
The cheese I got @SarawithanR lost its sq


To see the whole width of the table we make a small permanent change

In [225]:
pd.set_option('display.max_colwidth', None)

Let us see if some of the functions work properly and how we can use them to gain more information about this database. Here we list all the tweets that are generated by the user 'lost_dog'. We choose to only print the texts and nothing else.

In [226]:
print(df[df['user'].astype('string') == 'lost_dog']['text'])

43934              @NyleW I am lost. Please help me find a good home. 
45573             @SallyD I am lost. Please help me find a good home. 
46918         @zuppaholic I am lost. Please help me find a good home. 
47948         @LOSTPETUSA I am lost. Please help me find a good home. 
50571     @JeanLevertHood I am lost. Please help me find a good home. 
                                      ...                             
792408       @trooppetrie I am lost. Please help me find a good home. 
793313         @Carly_FTS I am lost. Please help me find a good home. 
793609         @inathlone I am lost. Please help me find a good home. 
798607              @Kram I am lost. Please help me find a good home. 
799404         @W_Hancock I am lost. Please help me find a good home. 
Name: text, Length: 549, dtype: object


Find the 50 users that generated the most tweets

In [227]:
df_grouped = df.groupby(['user']).count()
df_sorted = df_grouped.sort_values(by=['user'], ascending=False)
df_largest = df_sorted.nlargest(50, 'flag')
display(df_largest)

sentiment   id  date  flag  text
user                                             
lost_dog               549  549   549   549   549
webwoke                345  345   345   345   345
tweetpet               310  310   310   310   310
SallytheShizzle        281  281   281   281   281
VioletsCRUK            279  279   279   279   279
mcraddictal            276  276   276   276   276
tsarnick               248  248   248   248   248
what_bugs_u            246  246   246   246   246
Karen230683            238  238   238   238   238
DarkPiano              236  236   236   236   236
SongoftheOss           227  227   227   227   227
Jayme1988              225  225   225   225   225
keza34                 219  219   219   219   219
ramdomthoughts         216  216   216   216   216
shanajaca              213  213   213   213   213
wowlew                 212  212   212   212   212
nuttychris             211  211   211   211   211
TraceyHewins           211  211   211   211   211
thisgoeshere           207  207   207   207   207
Spidersamm             205  205   205   205   205
StDAY                  202  202   202   202   202
felicityfuller         195  195   195   195   195
Dogbook                192  192   192   192   192
_magic8ball            189  189   189   189   189
torilovesbradie        182  182   182   182   182
Dutchrudder            182  182   182   182   182
Djalfy                 182  182   182   182   182
twebbstack             180  180   180   180   180
Quimo                  180  180   180   180   180
enamoredsoul           179  179   179   179   179
Broooooke_             179  179   179   179   179
MTVnHollyWEST23        178  178   178   178   178
JessMcFlyxxx           178  178   178   178   178
MiDesfileNegro         177  177   177   177   177
linnetwoods            171  171   171   171   171
KevinEdwardsJr         171  171   171   171   171
insearchofnkotb        170  170   170   170   170
karinb_za              166  166   166   166   166
Scyranth               166  166   166   166   166
JBnVFCLover786         163  163   163   163   163
cookiemonster82        160  160   160   160   160
shellrawlins           159  159   159   159   159
maynaseric             159  159   159   159   159
hollyalyxfinch         159  159   159   159   159
mrs_mcsupergirl        158  158   158   158   158
lesley007              158  158   158   158   158
DonniesGirl69          155  155   155   155   155
paul_steele            152  152   152   152   152
judez_xo               152  152   152   152   152
patriciaco             151  151   151   151   151

<style>
  span {
    background-color: transparent;
    color: orange;
  }
</style>
What differs the tweets that share the same ID but are still saved in this dataframe, from the others, is that they have different sentiments; don't ask me why but some seem totally unreasonable. For instance, why should the following tweet, with the ID# 1467863684, be both positive and negative at the same time? Is it because they have mentioned the word "sad"?

<span>Awwh babs... you look so sad underneith that shop entrance of &quot;Yesterday's Musik&quot; O-: I like the look of the new transformer movie</span>

<h2>Investigating the Sentiments</h2>

As we see in Point * above, there are tweet IDs that are redundant. As a matter of fact, 1,598,314 out of 1,599,999 tweets have unique IDs. So, let us have a look at it and see what we can find out.

<h3>Step 1: Find Duplicates</h3>

Try and find how many tweets have more than one sentiment related to them. For this we count the occurrences of each value in the 'id' column

In [228]:
all_ids = df['id'].value_counts()
display(all_ids)

id
2190457769    2
1974742852    2
2062516845    2
1551586713    2
1563681287    2
             ..
2197311343    1
2197311196    1
2197311146    1
2197310899    1
2193602129    1
Name: count, Length: 1598314, dtype: int64

We then filter the result to include only values that appear twice or more

In [229]:
duplicated_tweet_ids = all_ids[all_ids >= 2]
display(duplicated_tweet_ids)

id
2190457769    2
1974742852    2
2062516845    2
1551586713    2
1563681287    2
             ..
2015412220    2
2006617256    2
1933201064    2
2189722020    2
1982279593    2
Name: count, Length: 1685, dtype: int64

This shows clearly that as we expected 1,599,999 - 1,598,314 = 1,685 tweets have double sentiments. So we print the rows in df for which the 'id' column is in repeated_values.index

In [230]:
duplicated_tweets = df[df['id'].isin(duplicated_tweet_ids.index)]
display(duplicated_tweets.sort_values(by='id').head(50))

sentiment          id                          date      flag   
212             0  1467863684  Mon Apr 06 22:33:35 PDT 2009  NO_QUERY  \
800260          4  1467863684  Mon Apr 06 22:33:35 PDT 2009  NO_QUERY   
274             0  1467880442  Mon Apr 06 22:38:04 PDT 2009  NO_QUERY   
800299          4  1467880442  Mon Apr 06 22:38:04 PDT 2009  NO_QUERY   
988             0  1468053611  Mon Apr 06 23:28:09 PDT 2009  NO_QUERY   
801279          4  1468053611  Mon Apr 06 23:28:09 PDT 2009  NO_QUERY   
1176            0  1468100580  Mon Apr 06 23:42:57 PDT 2009  NO_QUERY   
801572          4  1468100580  Mon Apr 06 23:42:57 PDT 2009  NO_QUERY   
1253            0  1468115720  Mon Apr 06 23:48:00 PDT 2009  NO_QUERY   
801649          4  1468115720  Mon Apr 06 23:48:00 PDT 2009  NO_QUERY   
1332            0  1468131748  Mon Apr 06 23:53:22 PDT 2009  NO_QUERY   
801793          4  1468131748  Mon Apr 06 23:53:22 PDT 2009  NO_QUERY   
1476            0  1468161883  Tue Apr 07 00:03:10 PDT 2009  NO_QUERY   
801960          4  1468161883  Tue Apr 07 00:03:10 PDT 2009  NO_QUERY   
802396          4  1468224250  Tue Apr 07 00:23:49 PDT 2009  NO_QUERY   
1751            0  1468224250  Tue Apr 07 00:23:49 PDT 2009  NO_QUERY   
802974          4  1468310350  Tue Apr 07 00:53:48 PDT 2009  NO_QUERY   
2126            0  1468310350  Tue Apr 07 00:53:48 PDT 2009  NO_QUERY   
803175          4  1468338634  Tue Apr 07 01:03:49 PDT 2009  NO_QUERY   
2251            0  1468338634  Tue Apr 07 01:03:49 PDT 2009  NO_QUERY   
803300          4  1468363676  Tue Apr 07 01:12:56 PDT 2009  NO_QUERY   
2374            0  1468363676  Tue Apr 07 01:12:56 PDT 2009  NO_QUERY   
3028            0  1468502040  Tue Apr 07 02:03:41 PDT 2009  NO_QUERY   
804315          4  1468502040  Tue Apr 07 02:03:41 PDT 2009  NO_QUERY   
3038            0  1468503801  Tue Apr 07 02:04:18 PDT 2009  NO_QUERY   
804348          4  1468503801  Tue Apr 07 02:04:18 PDT 2009  NO_QUERY   
3211            0  1468544973  Tue Apr 07 02:19:32 PDT 2009  NO_QUERY   
804655          4  1468544973  Tue Apr 07 02:19:32 PDT 2009  NO_QUERY   
3399            0  1468586841  Tue Apr 07 02:34:25 PDT 2009  NO_QUERY   
804928          4  1468586841  Tue Apr 07 02:34:25 PDT 2009  NO_QUERY   
805254          4  1468639063  Tue Apr 07 02:53:24 PDT 2009  NO_QUERY   
3631            0  1468639063  Tue Apr 07 02:53:24 PDT 2009  NO_QUERY   
3697            0  1468652839  Tue Apr 07 02:58:23 PDT 2009  NO_QUERY   
805340          4  1468652839  Tue Apr 07 02:58:23 PDT 2009  NO_QUERY   
805805          4  1468714181  Tue Apr 07 03:19:34 PDT 2009  NO_QUERY   
3970            0  1468714181  Tue Apr 07 03:19:34 PDT 2009  NO_QUERY   
4170            0  1468758512  Tue Apr 07 03:34:37 PDT 2009  NO_QUERY   
806090          4  1468758512  Tue Apr 07 03:34:37 PDT 2009  NO_QUERY   
806506          4  1468833927  Tue Apr 07 03:58:41 PDT 2009  NO_QUERY   
4479            0  1468833927  Tue Apr 07 03:58:41 PDT 2009  NO_QUERY   
5115            0  1468992236  Tue Apr 07 04:44:49 PDT 2009  NO_QUERY   
807441          4  1468992236  Tue Apr 07 04:44:49 PDT 2009  NO_QUERY   
5184            0  1469011145  Tue Apr 07 04:49:48 PDT 2009  NO_QUERY   
807542          4  1469011145  Tue Apr 07 04:49:48 PDT 2009  NO_QUERY   
6024            0  1469267615  Tue Apr 07 05:49:29 PDT 2009  NO_QUERY   
808644          4  1469267615  Tue Apr 07 05:49:29 PDT 2009  NO_QUERY   
809638          4  1469531660  Tue Apr 07 06:39:53 PDT 2009  NO_QUERY   
6729            0  1469531660  Tue Apr 07 06:39:53 PDT 2009  NO_QUERY   
809948          4  1469618724  Tue Apr 07 06:55:21 PDT 2009  NO_QUERY   
6946            0  1469618724  Tue Apr 07 06:55:21 PDT 2009  NO_QUERY   

                  user   
212           DjGundam  \
800260        DjGundam   
274            iCalvin   
800299         iCalvin   
988        mariejamora   
801279     mariejamora   
1176       cristygarza   
801572     cristygarza   
1253        WarholGirl   
801649      WarholGirl  

Now we check if those rows with duplicate ids have one 0 and one 4 in the sentiment column. We do it by adding the sentiments for each duplicated tweet. If they all have 4 as the result of the aggregation, we can conclude that each duplicate has received one 0 and one 4 as sentiment value. Consequenlty

In [231]:
aggregated_sentiments = duplicated_tweets.groupby('id').agg({'sentiment': 'sum', 'text': 'first'})
display(aggregated_sentiments)

sentiment   
id                      
1467863684          4  \
1467880442          4   
1468053611          4   
1468100580          4   
1468115720          4   
...               ...   
2193278017          4   
2193403830          4   
2193428118          4   
2193451289          4   
2193576442          4   

                                                                                                                                                      text  
id                                                                                                                                                          
1467863684  Awwh babs... you look so sad underneith that shop entrance of &quot;Yesterday's Musik&quot;  O-: I like the look of the new transformer movie   
1467880442                                                Haven't tweeted nearly all day  Posted my website tonight, hopefully that goes well  Night time!  
1468053611                  @hellobebe I also send some updates in plurk but i upload photos on twitter!  you didnt see any of my updates on plurk? Zero?   
1468100580                                                                        good night swetdreamss to everyonee   and jared never chat in kyte puff   
1468115720                     @ientje89 aw i'm fine too thanks! yeah i miss you so much on the MFC  but hope we can talk later on today  kisses :huglove:  
...                                                                                                                                                    ...  
2193278017                                             oh dear HH is back   please twitter do something about her.  I'm begging you, please pretty please   
2193403830                                                                english exam went okay        revising for french, r.e and geography now, urrff   
2193428118                                                                        finally finished typing!!!! Woohoooo  , still need to add graphs though   
2193451289                                           @fanafatin see, @misschimichanga tweet u to join us!! u really cant?  so if thurs, when &amp; where?   
2193576442                                      Had an injection today. Not fun  the rrst of the school day has been good. Tonight i am going dancing yey   

[1685 rows x 2 columns]

Which can be tested by asking if the number of 4s in the 'sentiment' column is the same as the size of the table

In [232]:
aggregated_sentiments['sentiment'].value_counts()

sentiment
4    1685
Name: count, dtype: int64

And this means <h3>YES</h3> We can go on with the next step and try to remove one of the duplicates and change the sentiment of the other one to 2, which is the same as 'natural'

<h3>Step 2: Remove Duplicates</h3>

Let us remove those rows in df that have the same id as in duplicated_tweet_ids and a sentiment of 0

In [233]:
df_wihout_duplicates = df[~(df['id'].isin(duplicated_tweet_ids.index) & (df['sentiment'] == 0))]
display(df_wihout_duplicates)

sentiment          id                          date      flag   
0                0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY  \
1                0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2                0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3                0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4                0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   
...            ...         ...                           ...       ...   
1599994          4  2193601966  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599995          4  2193601969  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599996          4  2193601991  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599997          4  2193602064  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599998          4  2193602129  Tue Jun 16 08:40:50 PDT 2009  NO_QUERY   

                    user   
0          scotthamilton  \
1               mattycus   
2                ElleCTF   
3                 Karoli   
4               joy_wolf   
...                  ...   
1599994  AmandaMarie1028   
1599995      TheWDBoards   
1599996           bpbabe   
1599997     tinydiamondz   
1599998   RyanTrevMorris   

                                                                                                                    text  
0        is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                              @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                        my whole body feels itchy and like its on fire   
3        @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                          @Kwesidei not the whole crew   
...                                                                                                                  ...  
1599994                                                         Just woke up. Having no school is the best feeling ever   
1599995                                   TheWDB.com - Very cool to hear old Walt interviews!  ♫ http://blip.fm/~8bmta  
1599996                                                        Are you ready for your MoJo Makeover? Ask me for details   
1599997                                                Happy 38th Birthday to my boo of alll time!!! Tupac Amaru Shakur   
1599998                                                   happy #charitytuesday @theNSPCC @SparksCharity @SpeakingUpH4H   

[1598314 rows x 6 columns]

<h3>Step 3: Neutralize the Sentiments</h3>

Set the sentiment of those rows in 'df' that are mentioned in the list of duplicated ids 'duplicated_tweet_ids' to 2 to mean that they should be classified - or percepted - as neutral

In [267]:
# change the sentiment of those rows in df_wihout_duplicates that are listed in duplicated_tweet_ids to 2
neutralized_df = df_wihout_duplicates.copy()
neutralized_df.loc[neutralized_df['id'].isin(duplicated_tweet_ids.index), 'sentiment'] = 2
display(neutralized_df)


sentiment          id                          date      flag   
0                0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY  \
1                0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2                0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3                0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4                0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   
...            ...         ...                           ...       ...   
1599994          4  2193601966  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599995          4  2193601969  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599996          4  2193601991  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599997          4  2193602064  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599998          4  2193602129  Tue Jun 16 08:40:50 PDT 2009  NO_QUERY   

                    user   
0          scotthamilton  \
1               mattycus   
2                ElleCTF   
3                 Karoli   
4               joy_wolf   
...                  ...   
1599994  AmandaMarie1028   
1599995      TheWDBoards   
1599996           bpbabe   
1599997     tinydiamondz   
1599998   RyanTrevMorris   

                                                                                                                    text  
0        is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                              @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                        my whole body feels itchy and like its on fire   
3        @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                          @Kwesidei not the whole crew   
...                                                                                                                  ...  
1599994                                                         Just woke up. Having no school is the best feeling ever   
1599995                                   TheWDB.com - Very cool to hear old Walt interviews!  ♫ http://blip.fm/~8bmta  
1599996                                                        Are you ready for your MoJo Makeover? Ask me for details   
1599997                                                Happy 38th Birthday to my boo of alll time!!! Tupac Amaru Shakur   
1599998                                                   happy #charitytuesday @theNSPCC @SparksCharity @SpeakingUpH4H   

[1598314 rows x 6 columns]

Let us check if the change has worked out

In [235]:
neutralized_df.apply(pd.Series.nunique)

sentiment          3
id           1598314
date          774362
flag               1
user          659775
text         1581465
dtype: int64

Yes indeed; we have three categories of sentiment now. And to be make sure

In [236]:
neutralized_df['sentiment'].value_counts()

sentiment
4    798315
0    798314
2      1685
Name: count, dtype: int64

We also skip the flag, that has no function in any types of investigations. Perhaps if we would combine it with other databases, we'll have to take it back. But for now, we just drop it.

In [268]:
neutralized_df = neutralized_df.drop(columns=['flag'])

And the result is

In [269]:
neutralized_df.head(50)

sentiment          id                          date             user   
0           0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton  \
1           0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
2           0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ElleCTF   
3           0  1467811193  Mon Apr 06 22:19:57 PDT 2009           Karoli   
4           0  1467811372  Mon Apr 06 22:20:00 PDT 2009         joy_wolf   
5           0  1467811592  Mon Apr 06 22:20:03 PDT 2009          mybirch   
6           0  1467811594  Mon Apr 06 22:20:03 PDT 2009             coZZ   
7           0  1467811795  Mon Apr 06 22:20:05 PDT 2009  2Hood4Hollywood   
8           0  1467812025  Mon Apr 06 22:20:09 PDT 2009          mimismo   
9           0  1467812416  Mon Apr 06 22:20:16 PDT 2009   erinx3leannexo   
10          0  1467812579  Mon Apr 06 22:20:17 PDT 2009     pardonlauren   
11          0  1467812723  Mon Apr 06 22:20:19 PDT 2009             TLeC   
12          0  1467812771  Mon Apr 06 22:20:19 PDT 2009  robrobbierobert   
13          0  1467812784  Mon Apr 06 22:20:20 PDT 2009      bayofwolves   
14          0  1467812799  Mon Apr 06 22:20:20 PDT 2009       HairByJess   
15          0  1467812964  Mon Apr 06 22:20:22 PDT 2009   lovesongwriter   
16          0  1467813137  Mon Apr 06 22:20:25 PDT 2009         armotley   
17          0  1467813579  Mon Apr 06 22:20:31 PDT 2009       starkissed   
18          0  1467813782  Mon Apr 06 22:20:34 PDT 2009        gi_gi_bee   
19          0  1467813985  Mon Apr 06 22:20:37 PDT 2009           quanvu   
20          0  1467813992  Mon Apr 06 22:20:38 PDT 2009       swinspeedx   
21          0  1467814119  Mon Apr 06 22:20:40 PDT 2009        cooliodoc   
22          0  1467814180  Mon Apr 06 22:20:40 PDT 2009       viJILLante   
23          0  1467814192  Mon Apr 06 22:20:41 PDT 2009       Ljelli3166   
24          0  1467814438  Mon Apr 06 22:20:44 PDT 2009    ChicagoCubbie   
25          0  1467814783  Mon Apr 06 22:20:50 PDT 2009      KatieAngell   
26          0  1467814883  Mon Apr 06 22:20:52 PDT 2009            gagoo   
27          0  1467815199  Mon Apr 06 22:20:56 PDT 2009          abel209   
28          0  1467815753  Mon Apr 06 22:21:04 PDT 2009  BaptisteTheFool   
29          0  1467815923  Mon Apr 06 22:21:07 PDT 2009        fatkat309   
30          0  1467815924  Mon Apr 06 22:21:07 PDT 2009            EmCDL   
31          0  1467815988  Mon Apr 06 22:21:09 PDT 2009         merisssa   
32          0  1467816149  Mon Apr 06 22:21:11 PDT 2009         Pbearfox   
33          0  1467816665  Mon Apr 06 22:21:21 PDT 2009             jsoo   
34          0  1467816749  Mon Apr 06 22:21:20 PDT 2009   scarletletterm   
35          0  1467817225  Mon Apr 06 22:21:27 PDT 2009      crosland_12   
36          0  1467817374  Mon Apr 06 22:21:30 PDT 2009          ajaxpro   
37          0  1467817502  Mon Apr 06 22:21:32 PDT 2009          Tmttq86   
38          0  1467818007  Mon Apr 06 22:21:39 PDT 2009   Anthony_Nguyen   
39          0  1467818020  Mon Apr 06 22:21:39 PDT 2009       itsanimesh   
40          0  1467818481  Mon Apr 06 22:21:46 PDT 2009        lionslamb   
41          0  1467818603  Mon Apr 06 22:21:49 PDT 2009        kennypham   
42          0  1467818900  Mon Apr 06 22:21:53 PDT 2009   DdubsShellBell   
43          0  1467819022  Mon Apr 06 22:21:56 PDT 2009      hpfangirl94   
44          0  1467819650  Mon Apr 06 22:22:05 PDT 2009        antzpantz   
45          0  1467819712  Mon Apr 06 22:22:06 PDT 2009        labrt2004   
46          0  1467819812  Mon Apr 06 22:22:07 PDT 2009        IrisJumbe   
47          0  1467820206  Mon Apr 06 22:22:13 PDT 2009         peacoats   
48          0  1467820835  Mon Apr 06 22:22:25 PDT 2009         cyantist   
49          0  1467820863  Mon Apr 06 22:22:23 PDT 2009           tautao   

                                                                                                                               

<style>span{background-color: transparent; color: orange;}</style>
We have 1,685 tweets with neutral sentiment. The amount of neutral sentiments is comparably very low, but still better nothing at all. <span>Thus, 'neutralized_df' is the table we are going to use for training our ML solution.</span>

hidden text: <!--now we need to gather all those having the same ID and the same everything but the sentiment, which has received both 0 and 4. Some are clearly ambiguous; others not. So what we do is that <span>we just remove the redundancy and what is remained will receive a sentiment of 2</span>. A little bit risky, since it still may be a source of confusion for the learner.-->

<h1>Side Track 1</h1>

Divide the 'date' into separate columns with the names 'day name', 'year', 'month', 'day', 'hour', 'minute', 'second'**

Let us see how many time zones are mentioned in the table

In [237]:
neutralized_df['date'].str.extract(r'([A-Z]{3})', expand=False).value_counts()

date
PDT    1598314
Name: count, dtype: int64

This shows that we only have PDT as the time zone. We can thus skip this by removing it from all the strings in the 'date' column

In [239]:
neutralized_zoneless_df = neutralized_df.copy()
neutralized_zoneless_df['date'] = neutralized_df['date'].str.replace('PDT', '')
neutralized_zoneless_df.head()

sentiment          id                       date      flag           user   
0          0  1467810672  Mon Apr 06 22:19:49  2009  NO_QUERY  scotthamilton  \
1          0  1467810917  Mon Apr 06 22:19:53  2009  NO_QUERY       mattycus   
2          0  1467811184  Mon Apr 06 22:19:57  2009  NO_QUERY        ElleCTF   
3          0  1467811193  Mon Apr 06 22:19:57  2009  NO_QUERY         Karoli   
4          0  1467811372  Mon Apr 06 22:20:00  2009  NO_QUERY       joy_wolf   

                                                                                                              text  
0  is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                        @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                  my whole body feels itchy and like its on fire   
3  @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                    @Kwesidei not the whole crew

Now we can divide the date up into 6 different columns, for other investigations than the main subject of this project

In [266]:
neutralized_zoneless_separated_df = neutralized_zoneless_df.copy()
# Convert the 'date' column to a datetime type
neutralized_zoneless_separated_df['date'] = pd.to_datetime(neutralized_zoneless_separated_df['date'], format='%a %b %d %H:%M:%S %Y')
# Extract individual date components into separate columns
neutralized_zoneless_separated_df['year'] = neutralized_zoneless_separated_df['date'].dt.year
neutralized_zoneless_separated_df['month'] = neutralized_zoneless_separated_df['date'].dt.month
neutralized_zoneless_separated_df['day'] = neutralized_zoneless_separated_df['date'].dt.day
neutralized_zoneless_separated_df['hour'] = neutralized_zoneless_separated_df['date'].dt.hour
neutralized_zoneless_separated_df['minute'] = neutralized_zoneless_separated_df['date'].dt.minute
neutralized_zoneless_separated_df['second'] = neutralized_zoneless_separated_df['date'].dt.second
neutralized_zoneless_separated_df['weekday'] = neutralized_zoneless_separated_df['date'].dt.weekday
# Drop the original 'date' column
neutralized_zoneless_separated_df = neutralized_zoneless_separated_df.drop('date', axis=1)

Now show the neutralized, zoneless df with separated date details

In [265]:
neutralized_zoneless_separated_df.sort_values(by='weekday', ascending=True).head(1599999)

sentiment          id      flag           user   
0                0  1467810672  NO_QUERY  scotthamilton  \
753137           0  2286778838  NO_QUERY   lovingjordan   
753136           0  2286778725  NO_QUERY       RenaWaka   
753135           0  2286778561  NO_QUERY        Bgarzza   
753134           0  2286778000  NO_QUERY       sega_123   
...            ...         ...       ...            ...   
879353           4  1685426805  NO_QUERY       zaibatsu   
879354           4  1685426828  NO_QUERY        Kbear_3   
879355           4  1685426956  NO_QUERY       jaded0ll   
879346           4  1685426618  NO_QUERY      ongmaggie   
1161604          4  1979521342  NO_QUERY  x_rachybear_x   

                                                                                                                                              text   
0                                  is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  \
753137                                                                                   I am so sad about Jon and Kate n their beautiful kids...    
753136                                                                                                          when is the rain going to Stop?!!    
753135                                                                                                             I need some more chap stick...    
753134                                                  jeanette is sick. and taking a nap.  but her brother is flying over to cali this tuesday.    
...                                                                                                                                            ...   
879353                                                                                                  @rosehwang It's not my puppy   Doing great   
879354                                                                                                                             He's so sweeet    
879355   @demois i'm sorry i wasn't able to be one of your plurk defenders. lol. but i promise to help you next time someone b*tches on you. lol.    
879346                                                                       @oyebto nope, it was at 11+pm. giant at tampines opens till midnight    
1161604                                                                                         the sleepover was fun  and now im eating maltesers   

         year  month  day  hour  minute  second  weekday  
0        2009      4    6    22      19      49        0  
753137   2009      6   22    16      54      52        0  
753136   2009      6   22    16      54      52        0  
753135   2009      6   22    16      54      51        0  
753134   2009      6   22    16      54      48        0  
...       ...    ...  ...   ...     ...     ...      ...  
879353   2009      5    3     0      20      31        6  
879354   2009      5    3     0      20      31        6  
879355   2009      5    3     0      20      34        6  
879346   2009      5    3     0      20      28        6  
1161604  2009      5   31     3      24      30        6  

[1598314 rows x 12 columns]

In [ ]:
#the previous way copilot suggested for spliting the date column into six separate columns with the titles year, month, day, hour, minute, second
#neutralized_zoneless_separated_df[['year', 'month', 'day', 'hour', 'minute', 'second']] = neutralized_zoneless_separated_df['date'].str.split(' ', expand=True)
#not the above way copilot suggested for spliting the date column into six separate columns with the titles year, month, day, hour, minute, second
#no not this way... copilot suggested some other way before this one

#
# divide the data column into six separate columns with the titles year, month, day, hour, minute, second using format


And the result looks like what follows

In [247]:
display(neutralized_zoneless_separated_df.head())

sentiment          id                       date      flag           user   
0          0  1467810672  Mon Apr 06 22:19:49  2009  NO_QUERY  scotthamilton  \
1          0  1467810917  Mon Apr 06 22:19:53  2009  NO_QUERY       mattycus   
2          0  1467811184  Mon Apr 06 22:19:57  2009  NO_QUERY        ElleCTF   
3          0  1467811193  Mon Apr 06 22:19:57  2009  NO_QUERY         Karoli   
4          0  1467811372  Mon Apr 06 22:20:00  2009  NO_QUERY       joy_wolf   

                                                                                                              text   
0  is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  \
1                        @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds   
2                                                                  my whole body feels itchy and like its on fire    
3  @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.    
4                                                                                    @Kwesidei not the whole crew    

  year month day      hour minute second  
0  Mon   Apr  06  22:19:49          2009  
1  Mon   Apr  06  22:19:53          2009  
2  Mon   Apr  06  22:19:57          2009  
3  Mon   Apr  06  22:19:57          2009  
4  Mon   Apr  06  22:20:00          2009

Show the result, sorted by time zone in descending order

In [ ]:
df.sort_values(by='time zone', ascending=False).head(10)

KeyError: 'time zone'

**TODO : Find the ten dates when the most tweets were generated**

In [ ]:
doubleID_df = df.groupby('id').count() > 1
double_id = df.loc[df['id'].isin(doubleID_df.loc[doubleID_df['sentiment'] == True].index)]
double_id.sort_values(by='id').head(100)

**TODO : Find out the change in the frequnency of tweets per day**

**TODO : See if there is a correlation between the sentiment of the tweet and the day it was made**

**TODO : See if there is a correlation between the sentiment of the tweet and the time of the day it was made; eg if the tweets tens to be more negative at night compared to days and so on**

In [ ]:
double_sentiment_id = df.groupby(['id']).count().sort_values(by='sentiment', ascending=False)
double_sentiment_id
#double_sentiment_id.loc[double_sentiment_id['sentiment'] > 1]
#df.groupby(['id']).count().sort_values(by='sentiment', ascending=False)

sentiment  date  flag  user  text
id                                           
2013708350          2     2     2     2     2
1694607515          2     2     2     2     2
1981455916          2     2     2     2     2
1977625382          2     2     2     2     2
1557375700          2     2     2     2     2
...               ...   ...   ...   ...   ...
1974380475          1     1     1     1     1
1974380466          1     1     1     1     1
1974380370          1     1     1     1     1
1974380271          1     1     1     1     1
2329205794          1     1     1     1     1

[1598314 rows x 5 columns]

In [ ]:
doubleID_df = df.groupby('id').count() > 1
double_id = df.loc[df['id'].isin(doubleID_df.loc[doubleID_df['sentiment'] == True].index)]
double_id.sort_values(by='id').head(100)

sentiment          id                          date      flag   
212             0  1467863684  Mon Apr 06 22:33:35 PDT 2009  NO_QUERY  \
800260          4  1467863684  Mon Apr 06 22:33:35 PDT 2009  NO_QUERY   
274             0  1467880442  Mon Apr 06 22:38:04 PDT 2009  NO_QUERY   
800299          4  1467880442  Mon Apr 06 22:38:04 PDT 2009  NO_QUERY   
988             0  1468053611  Mon Apr 06 23:28:09 PDT 2009  NO_QUERY   
...           ...         ...                           ...       ...   
824538          4  1556267135  Sat Apr 18 22:12:27 PDT 2009  NO_QUERY   
17812           0  1556402800  Sat Apr 18 22:42:07 PDT 2009  NO_QUERY   
825096          4  1556402800  Sat Apr 18 22:42:07 PDT 2009  NO_QUERY   
825935          4  1556571114  Sat Apr 18 23:22:47 PDT 2009  NO_QUERY   
18340           0  1556571114  Sat Apr 18 23:22:47 PDT 2009  NO_QUERY   

               user   
212        DjGundam  \
800260     DjGundam   
274         iCalvin   
800299      iCalvin   
988     mariejamora   
...             ...   
824538      banerch   
17812   StormClaudi   
825096  StormClaudi   
825935     Kikirowr   
18340      Kikirowr   

                                                                                                                                                  text  
212     Awwh babs... you look so sad underneith that shop entrance of &quot;Yesterday's Musik&quot;  O-: I like the look of the new transformer movie   
800260  Awwh babs... you look so sad underneith that shop entrance of &quot;Yesterday's Musik&quot;  O-: I like the look of the new transformer movie   
274                                                   Haven't tweeted nearly all day  Posted my website tonight, hopefully that goes well  Night time!  
800299                                                Haven't tweeted nearly all day  Posted my website tonight, hopefully that goes well  Night time!  
988                     @hellobebe I also send some updates in plurk but i upload photos on twitter!  you didnt see any of my updates on plurk? Zero?   
...                                                                                                                                                ...  
824538                                 @meladiaz HAHAHAHAHA! ) i miss you toooooo!  i have kwentooo!  i'll tell you next time we see each other. :&gt;  
17812               @IzzyJ_Is_Here haha yayyy! aw, lady i really hope you're not sad. that's no fun  just listen to calle ocho and dance dance dance!   
825096              @IzzyJ_Is_Here haha yayyy! aw, lady i really hope you're not sad. that's no fun  just listen to calle ocho and dance dance dance!   
825935                This little tree is tiiiiired  25's (and dealing with stupid people) tomorrow and then finishing 10s monday! Raiding every nice   
18340                 This little tree is tiiiiired  25's (and dealing with stupid people) tomorrow and then finishing 10s monday! Raiding every nice   

[100 rows x 6 columns]

In [ ]:
df.groupby('id').count()>1

sentiment   date   flag   user   text
id                                               
1467810672      False  False  False  False  False
1467810917      False  False  False  False  False
1467811184      False  False  False  False  False
1467811193      False  False  False  False  False
1467811372      False  False  False  False  False
...               ...    ...    ...    ...    ...
2329205009      False  False  False  False  False
2329205038      False  False  False  False  False
2329205473      False  False  False  False  False
2329205574      False  False  False  False  False
2329205794      False  False  False  False  False

[1598314 rows x 5 columns]

In [ ]:
df.groupby('id')['id'].count()>1

id
1467810672    False
1467810917    False
1467811184    False
1467811193    False
1467811372    False
              ...  
2329205009    False
2329205038    False
2329205473    False
2329205574    False
2329205794    False
Name: id, Length: 1598314, dtype: bool

In [ ]:
pd.set_option('display.max_rows', 10000)

In [ ]:
df

sentiment          id                          date      flag   
0                0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY  \
1                0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2                0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3                0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4                0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   
...            ...         ...                           ...       ...   
1599994          4  2193601966  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599995          4  2193601969  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599996          4  2193601991  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599997          4  2193602064  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599998          4  2193602129  Tue Jun 16 08:40:50 PDT 2009  NO_QUERY   

                    user   
0          scotthamilton  \
1               mattycus   
2                ElleCTF   
3                 Karoli   
4               joy_wolf   
...                  ...   
1599994  AmandaMarie1028   
1599995      TheWDBoards   
1599996           bpbabe   
1599997     tinydiamondz   
1599998   RyanTrevMorris   

                                                                                                                    text  
0        is upset that he can't update his Facebook by texting it... and might cry as a result  School today also. Blah!  
1                              @Kenichan I dived many times for the ball. Managed to save 50%  The rest go out of bounds  
2                                                                        my whole body feels itchy and like its on fire   
3        @nationwideclass no, it's not behaving at all. i'm mad. why am i here? because I can't see you all over there.   
4                                                                                          @Kwesidei not the whole crew   
...                                                                                                                  ...  
1599994                                                         Just woke up. Having no school is the best feeling ever   
1599995                                   TheWDB.com - Very cool to hear old Walt interviews!  ♫ http://blip.fm/~8bmta  
1599996                                                        Are you ready for your MoJo Makeover? Ask me for details   
1599997                                                Happy 38th Birthday to my boo of alll time!!! Tupac Amaru Shakur   
1599998                                                   happy #charitytuesday @theNSPCC @SparksCharity @SpeakingUpH4H   

[1599999 rows x 6 columns]

In [ ]:
# extract the row with the id number 1467811372
df.loc[df['id'] == 1467813782]

sentiment          id                          date      flag       user   
18          0  1467813782  Mon Apr 06 22:20:34 PDT 2009  NO_QUERY  gi_gi_bee  \

                                                                               text  
18  @FakerPattyPattz Oh dear. Were you drinking out of the forgotten table drinks?

In [ ]:
# create a sample dataframe
daf = pd.DataFrame({'col1': [1, 2, 2, 3, 3, 3, 4, 4, 4, 4]})
display(daf)
# count the occurrences of each value in the 'col1' column
value_counts = daf['col1'].value_counts()
display(value_counts)
# filter the result to include only values that appear three times or more
repeated_values = value_counts[value_counts >= 3]
display(repeated_values)
# print the repeated values
print(repeated_values.index.tolist())

col1
0     1
1     2
2     2
3     3
4     3
5     3
6     4
7     4
8     4
9     4

col1
4    4
3    3
2    2
1    1
Name: count, dtype: int64

col1
4    4
3    3
Name: count, dtype: int64

[4, 3]
